File: apps/frontend/lib/types/evaluation.ts
/**
 * Types for the model evaluation dashboard (issue #79).
 *
 * Contract shared with the backend: apps/backend/app/schemas/evaluation.py
 * mirrors these interfaces field-for-field. Change both together.
 */

export interface PerClassMetrics {
  precision: number
  recall: number
  f1: number
  support: number
}

export interface ClassificationMetrics {
  accuracy: number
  precision_macro: number
  precision_weighted: number
  recall_macro: number
  recall_weighted: number
  f1_macro: number
  f1_weighted: number
  roc_auc: number | null
  log_loss: number | null
  per_class_metrics: Record<string, PerClassMetrics>
}

export interface RegressionMetrics {
  mae: number
  mse: number
  rmse: number
  r2: number
  mape: number | null
}

export interface ConfusionMatrixData {
  labels: string[]
  /** matrix[i][j] = count of actual labels[i] predicted as labels[j] */
  matrix: number[][]
}

export interface CurvePoint {
  x: number
  y: number
  threshold?: number | null
}

export interface ROCCurveData {
  /** Per-class ROC curves (one-vs-rest for multiclass). x=FPR, y=TPR. */
  curves: Record<string, CurvePoint[]>
  auc_per_class: Record<string, number>
  macro_auc: number | null
}

export interface PRCurveData {
  /** Per-class precision-recall curves. x=recall, y=precision. */
  curves: Record<string, CurvePoint[]>
  /** Positive-class prevalence per class (random-classifier baseline) */
  baseline_per_class: Record<string, number>
}

export interface AIExplanation {
  overall_assessment: string
  metric_explanations: Record<string, string>
  strengths: string[]
  concerns: string[]
  recommendations: string[]
  generated_by: 'openai' | 'fallback'
}

export interface ModelEvaluationResponse {
  model_id: string
  model_name: string | null
  algorithm: string | null
  problem_type: string
  /**
   * True when evaluation artifacts are unavailable (models trained before
   * issue #79) and only stored scalar metrics are returned.
   */
  partial: boolean
  /**
   * True when these metrics were computed on the calibrator's own fit data
   * rather than a clean held-out set, so they may be optimistically biased
   * (issue #201 small-data fallback). False on the honest split.
   */
  evaluation_on_calibration_set: boolean
  metrics: ClassificationMetrics | RegressionMetrics | null
  /** Scalar metrics persisted at training time (cv_score, test_score, ...) */
  stored_metrics: Record<string, number>
  confusion_matrix: ConfusionMatrixData | null
  roc_curve: ROCCurveData | null
  pr_curve: PRCurveData | null
  feature_importance: Record<string, number> | null
  ai_explanation: AIExplanation | null
  evaluated_at: string
}

/** One feature's importance, for a ranked importance list (issue #80). */
export interface RankedFeature {
  feature_name: string
  importance: number
}

/**
 * Global feature importance for one model (issue #80). Carries model-native
 * importance and, when the model type is SHAP-supported, SHAP-based importance.
 */
export interface FeatureImportanceResponse {
  model_id: string
  partial: boolean
  /** "tree", "linear", or null when SHAP is unavailable */
  explainer_type: string | null
  native_importance: RankedFeature[]
  shap_importance: RankedFeature[] | null
  message: string | null
}

/**
 * SHAP summary-plot data for one model (issue #80). feature_importance is the
 * mean |SHAP| per feature, ranked. partial=true (with empty lists + a message)
 * for models trained before #80 or whose type isn't SHAP-supported.
 */
export interface ShapSummaryResponse {
  model_id: string
  partial: boolean
  explainer_type: string | null
  problem_type: string
  feature_importance: RankedFeature[]
  base_value: number | null
  plain_language: string
  message: string | null
  evaluated_at: string
}

export interface ModelComparisonRequest {
  model_ids: string[]
}

export interface ModelEvaluationSummary {
  model_id: string
  name: string
  algorithm: string
  problem_type: string
  cv_score: number | null
  test_score: number | null
  metrics: Record<string, number>
  created_at: string | null
}

export interface ModelComparisonResponse {
  problem_type: string
  dataset_id: string
  models: ModelEvaluationSummary[]
}

// Model versioning & history (issue #78). Mirrors app/schemas/evaluation.py.

export interface ModelVersionEntry {
  model_id: string
  version_number: number
  name: string
  algorithm: string
  problem_type: string
  cv_score: number | null
  test_score: number | null
  is_production: boolean
  is_active: boolean
  created_at: string | null
  promoted_at: string | null
  dataset_id: string
  dataset_version_id: string | null
  parent_model_id: string | null
  feature_names: string[]
  environment_metadata: Record<string, unknown> | null
  version_notes: string | null
}

export interface ModelVersionListResponse {
  model_id: string
  dataset_id: string
  name: string
  total: number
  production_model_id: string | null
  versions: ModelVersionEntry[]
}

export interface PromoteVersionResponse {
  model_id: string
  is_production: boolean
  promoted_at: string | null
  demoted_model_ids: string[]
}

/** Type guard: classification vs regression metrics. */
export function isClassificationMetrics(
  metrics: ClassificationMetrics | RegressionMetrics
): metrics is ClassificationMetrics {
  return 'accuracy' in metrics
}

// --- Error analysis (issue #81) — mirrors app/schemas/error_analysis.py ---

export interface ErrorDistribution {
  total_samples: number
  total_errors: number
  overall_error_rate: number
  per_class_error_rate: Record<string, number>
}

export interface ConfusionPair {
  actual: string
  predicted: string
  count: number
  rate: number
}

export interface ErrorSegment {
  feature: string
  range_label: string
  lower: number | null
  upper: number | null
  error_rate: number
  error_count: number
  sample_count: number
}

export interface ErrorCluster {
  cluster_id: number
  size: number
  characteristics: string[]
  dominant_confusion: string | null
}

export interface ErrorPattern {
  rule: string
  error_rate: number
  error_count: number
  sample_count: number
}

export interface ErrorCase {
  index: number
  actual: string
  predicted: string
  confidence: number | null
  top_features: Record<string, number>
}

export interface ErrorAnalysisResponse {
  model_id: string
  model_name: string | null
  algorithm: string | null
  problem_type: string
  partial: boolean
  distribution: ErrorDistribution | null
  confusion_pairs: ConfusionPair[]
  segments: ErrorSegment[]
  clusters: ErrorCluster[]
  patterns: ErrorPattern[]
  cases: ErrorCase[]
  suggestions: string[]
  suggestions_generated_by: 'openai' | 'fallback'
  message: string | null
  evaluated_at: string
}

File: apps/backend/app/schemas/evaluation.py
"""Schemas for the model evaluation dashboard (issue #79).

Contract shared with the frontend: apps/frontend/lib/types/evaluation.ts
mirrors these models field-for-field. Change both together.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PerClassMetrics(BaseModel):
    """Precision/recall/F1 for a single class."""

    precision: float
    recall: float
    f1: float
    support: int = Field(..., ge=0, description="Number of true samples of this class")


class ClassificationMetrics(BaseModel):
    """Aggregate classification metrics computed on the held-out test set."""

    accuracy: float
    precision_macro: float
    precision_weighted: float
    recall_macro: float
    recall_weighted: float
    f1_macro: float
    f1_weighted: float
    roc_auc: float | None = Field(
        None, description="Macro one-vs-rest AUC; None when probabilities are unavailable"
    )
    log_loss: float | None = Field(
        None, description="None when probabilities are unavailable"
    )
    per_class_metrics: dict[str, PerClassMetrics] = Field(default_factory=dict)


class RegressionMetrics(BaseModel):
    """Aggregate regression metrics computed on the held-out test set."""

    mae: float
    mse: float
    rmse: float
    r2: float
    mape: float | None = Field(
        None, description="None when y_test contains zeros (undefined)"
    )


class ConfusionMatrixData(BaseModel):
    """Confusion matrix; matrix[i][j] = count of actual labels[i] predicted as labels[j]."""

    labels: list[str]
    matrix: list[list[int]]


class CurvePoint(BaseModel):
    """One point on a ROC or PR curve."""

    x: float
    y: float
    threshold: float | None = None


class ROCCurveData(BaseModel):
    """Per-class ROC curves (one-vs-rest for multiclass). x=FPR, y=TPR."""

    curves: dict[str, list[CurvePoint]] = Field(default_factory=dict)
    auc_per_class: dict[str, float] = Field(default_factory=dict)
    macro_auc: float | None = None


class PRCurveData(BaseModel):
    """Per-class precision-recall curves. x=recall, y=precision."""

    curves: dict[str, list[CurvePoint]] = Field(default_factory=dict)
    baseline_per_class: dict[str, float] = Field(
        default_factory=dict,
        description="Positive-class prevalence per class (random-classifier baseline)",
    )


class AIExplanation(BaseModel):
    """Plain-language 'Model Report Card' explanation."""

    overall_assessment: str
    metric_explanations: dict[str, str] = Field(default_factory=dict)
    strengths: list[str] = Field(default_factory=list)
    concerns: list[str] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)
    generated_by: Literal["openai", "fallback"] = "fallback"


class ModelEvaluationResponse(BaseModel):
    """Full evaluation payload for one model."""

    model_id: str
    model_name: str | None = None
    algorithm: str | None = None
    problem_type: str
    partial: bool = Field(
        False,
        description=(
            "True when evaluation artifacts are unavailable (models trained before "
            "issue #79) and only stored scalar metrics are returned"
        ),
    )
    metrics: ClassificationMetrics | RegressionMetrics | None = None
    stored_metrics: dict[str, float] = Field(
        default_factory=dict,
        description="Scalar metrics persisted at training time (cv_score, test_score, ...)",
    )
    confusion_matrix: ConfusionMatrixData | None = None
    roc_curve: ROCCurveData | None = None
    pr_curve: PRCurveData | None = None
    feature_importance: dict[str, float] | None = None
    ai_explanation: AIExplanation | None = None
    evaluated_at: datetime


class RankedFeature(BaseModel):
    """One feature's importance, for a ranked importance list (issue #80)."""

    feature_name: str
    importance: float


class FeatureImportanceResponse(BaseModel):
    """Global feature importance for one model (issue #80).

    Carries both the model-native importance (always available when the model
    exposed coef_/feature_importances_/SelectKBest at training) and, when the
    model type is SHAP-supported, SHAP-based importance. ``partial`` is True
    only when neither is available.
    """

    model_id: str
    partial: bool = False
    explainer_type: str | None = Field(
        None, description='"tree", "linear", or None when SHAP is unavailable'
    )
    native_importance: list[RankedFeature] = Field(default_factory=list)
    shap_importance: list[RankedFeature] | None = None
    message: str | None = None


class ShapSummaryResponse(BaseModel):
    """SHAP summary-plot data for one model (issue #80).

    ``feature_importance`` is the mean |SHAP| per feature, ranked. For models
    trained before #80 or whose type isn't SHAP-supported, ``partial`` is True,
    the lists are empty, and ``message`` explains the fallback to native
    importance — the endpoint never 500s for an owned model.
    """

    model_id: str
    partial: bool = False
    explainer_type: str | None = None
    problem_type: str
    feature_importance: list[RankedFeature] = Field(default_factory=list)
    base_value: float | None = None
    plain_language: str = ""
    message: str | None = None
    evaluated_at: datetime


class ModelComparisonRequest(BaseModel):
    """Request body for POST /api/v1/ml/compare."""

    model_ids: list[str] = Field(..., min_length=2, max_length=5)


class ModelEvaluationSummary(BaseModel):
    """One model's row in a comparison."""

    model_id: str
    name: str
    algorithm: str
    problem_type: str
    cv_score: float | None = None
    test_score: float | None = None
    metrics: dict[str, float] = Field(default_factory=dict)
    created_at: datetime | None = None


class ModelComparisonResponse(BaseModel):
    """Response for POST /api/v1/ml/compare."""

    problem_type: str
    dataset_id: str
    models: list[ModelEvaluationSummary]
